feat(group): Add 'votingDuration' field and implement its effects in …

daniel-herrero opened the pull request:
This MR adds `votingDuration` field to `Group`, reflecting the number of hours a voting will be open to vote.
The application should be updated consequently.

https://tree.taiga.io/project/pabloruiz-kaleidos-patio/us/68?no-milestone=1

![gif](https://media.giphy.com/media/QdbIz3R4GjXAk/giphy.gif)

### Validation Tests

**Commits**
- [ ] Review the code

**Database**
- [ ] Schema
- [ ] Fixtures
- [ ] Migrations

**API operations**
- [ ] The new field should be publicly exposed in the API
- [ ] Creating/updating a group should consider the new field (and default values when omitted - 24 hours)

**Votings / Votes**
The new concept _voting period_ arises (`votingTime` + `votingDuration`), and it's critical to pay special attention to:
- [ ] The creation of new votings/notifications when a new voting period begins
- [ ] Existing votings should be expired as soon as It's detected that a voting it's out of its voting period
- [ ] That votes are legal while created during the voting period (or illegal and forbidden in any other case)

mariogarcia commented on src/main/java/patio/group/domain/Group.java:
```diff
@@ -64,6 +64,9 @@
   private Set<UserGroup> users;
 
+  @Column(name = "voting_duration")
+  private Integer votingDuration; /* in hours */
```
Are you sure you only want hours ? This avoid setting minutes

daniel-herrero replied:
I'll double check it wiht Miryam...

daniel-herrero replied:
She has confirm that we don't intent to manage minutes, but it worth it to give her the chance. :+1:

mariogarcia commented:
Remember previously migrations should be removed at some point, because in order to work we skipped migrations checksum. Flyway should check the signature of every migration.

daniel-herrero commented:
> Remember previously migrations should be removed at some point, because in order to work we skipped migrations checksum. Flyway should check the signature of every migration.

I'm not sure If I'm understanding you fully. What are you suggesting?.
Is it a reminder to re-install the signature check ASAP? Do I have to fix something related to migrations in this MR?

We had to remove the signature check because of constraints in PRO's database, but the branch is not merged yet there (and won't be in a time). So, regarding migrations, I guess we should remain the same until then.

daniel-herrero commented:
> I can't test this right now, so I assume everything is passing the tests and you've test it in runtime.

Yes, I have test it locally several times in runtime, and loading different situations in the fixtures. It seems to work properly, closing the old votings and creating the new ones when its time arrives (voting_time + voting_duration).

My main concern was the Time Zones, but the `voting_time` I used in the fixtures were loaded that way (f.i.: `time with time zone '12:48:00.146512+01:00` and everything appears to behave as it should. So I guess it should work well.

Do you have in mind another test to perform?

mariogarcia commented:
> > Remember previously migrations should be removed at some point, because in order to work we skipped migrations checksum. Flyway should check the signature of every migration.
> 
> I'm not sure If I'm understanding you fully. What are you suggesting?.
> Is it a reminder to re-install the signature check ASAP? Do I have to fix something related to migrations in this MR?
> 
> We had to remove the signature check because of constraints in PRO's database, but the branch is not merged yet there (and won't be in a time). So, regarding migrations, I guess we should remain the same until then.

We had to remove it because we changed file HEADERS. And because migration headers changed their signature changed too. That caused migration checksum to fail. This MUST NOT BE ADDRESSED in this MR but it has to be addressed soon. And also it has to be addressed in development. Once migrations have been fixed in development I'd go and merge development to master and create tag 1.0.0 in api.

daniel-herrero commented:
> > > Remember previously migrations should be removed at some point, because in order to work we skipped migrations checksum. Flyway should check the signature of every migration.
> > 
> > 
> > I'm not sure If I'm understanding you fully. What are you suggesting?.
> > Is it a reminder to re-install the signature check ASAP? Do I have to fix something related to migrations in this MR?
> > We had to remove the signature check because of constraints in PRO's database, but the branch is not merged yet there (and won't be in a time). So, regarding migrations, I guess we should remain the same until then.
> 
> We had to remove it because we changed file HEADERS. And because migration headers changed their signature changed too. That caused migration checksum to fail. This MUST NOT BE ADDRESSED in this MR but it has to be addressed soon. And also it has to be addressed in development. Once migrations have been fixed in development I'd go and merge development to master and create tag 1.0.0 in api.

Ok, let me try to explain how I see things right now in order to verify we're on the same page. Please, correct me at any point:

1. We detected a problem in the migrations currently deployed in Kaleidos (in the production environment).
2. The checksum stored in its database (`flyway_schema_history.checksum field`) would be different to the ones that were about to being upload, so the rest of migrations woudn't be applied creating an inconsistent status we wanted to avoid.
3. For this reason we introduced a temporal change in the configuration, to skip the checksum valdidation. Involved commit: https://github.com/patio-team/patio-api/pull/98/commits/6dd92cab5a8dad03a5667f47bd55e73cf828716d
 `flyway.datasources.default.validate-on-migrate: false`
4. This change were introduced to both development and production environments (application-template.yml - application.yml / application-prod-yml)
5. We should get back to normal ASAP, reverting the changes introduced in that moment to both environments (dev/pro)

As I see it, I think we can't remove the flyway change (at least in application-prod.yml) until we DEPLOY changes into Kaleidos's SERVERS and its database, or we would still face the sane problem with migrations.

And the main issues against DEPLOYING in Kaleidos are:
* To avoid deployment risks
* There are potentially dangerous changes (as the change of DWBH in db and config files, ..., or the refactoring of the votings's timings in this MR)
* There are changes in back that wouldn't be yet enjoyed by front (frontend-pending)
* I'm the only one responsible here right now (of the deployment, and to solve any potential issue that may arise with the changes)

So, long story short, we should keep the config change in flyway ('application-prod.yml' file) until we DEPLOY in Kaleidos. We can remove the changes that affect just to the `development environment`, fix current deployment in Okteto, and merge it labeling into master at any point, but noy updating Kaleidos's servers.

Presently I'm the only one that keeps working in the project, and my intention were to keep introducing new backend functionalities without harming frontend altought it becomes outdated. And taking into account the situation, I would keep the flyway changes as they're right now (or if you prefer we can remove it from application-template.yml), creating an issue in Taiga/Github to avoid forgeting to revert the change in the production environment.

Sorry for the extension of the comment. I'm available to hold a remote meeting at any time.

daniel-herrero commented:
> > > > Remember previously migrations should be removed at some point, because in order to work we skipped migrations checksum. Flyway should check the signature of every migration.
> > > 
> > > 
> > > I'm not sure If I'm understanding you fully. What are you suggesting?.
> > > Is it a reminder to re-install the signature check ASAP? Do I have to fix something related to migrations in this MR?
> > > We had to remove the signature check because of constraints in PRO's database, but the branch is not merged yet there (and won't be in a time). So, regarding migrations, I guess we should remain the same until then.
> > 
> > 
> > We had to remove it because we changed file HEADERS. And because migration headers changed their signature changed too. That caused migration checksum to fail. This MUST NOT BE ADDRESSED in this MR but it has to be addressed soon. And also it has to be addressed in development. Once migrations have been fixed in development I'd go and merge development to master and create tag 1.0.0 in api.
> 
> Ok, let me try to explain how I see things right now in order to verify we're on the same page. Please, correct me at any point:
> 
> 1. We detected a problem in the migrations currently deployed in Kaleidos (in the production environment).
> 2. The checksum stored in its database (`flyway_schema_history.checksum field`) would be different to the ones that were about to being upload, so the rest of migrations woudn't be applied creating an inconsistent status we wanted to avoid.
> 3. For this reason we introduced a temporal change in the configuration, to skip the checksum valdidation. Involved commit: [6dd92ca](https://github.com/patio-team/patio-api/commit/6dd92cab5a8dad03a5667f47bd55e73cf828716d)
>    `flyway.datasources.default.validate-on-migrate: false`
> 4. This change were introduced to both development and production environments (application-template.yml - application.yml / application-prod-yml)
> 5. We should get back to normal ASAP, reverting the changes introduced in that moment to both environments (dev/pro)
> 
> As I see it, I think we can't remove the flyway change (at least in application-prod.yml) until we DEPLOY changes into Kaleidos's SERVERS and its database, or we would still face the sane problem with migrations.
> 
> And the main issues against DEPLOYING in Kaleidos are:
> 
> * To avoid deployment risks
> * There are potentially dangerous changes (as the change of DWBH in db and config files, ..., or the refactoring of the votings's timings in this MR)
> * There are changes in back that wouldn't be yet enjoyed by front (frontend-pending)
> * I'm the only one responsible here right now (of the deployment, and to solve any potential issue that may arise with the changes)
> 
> So, long story short, we should keep the config change in flyway ('application-prod.yml' file) until we DEPLOY in Kaleidos. We can remove the changes that affect just to the `development environment`, fix current deployment in Okteto, and merge it labeling into master at any point, but noy updating Kaleidos's servers.
> 
> Presently I'm the only one that keeps working in the project, and my intention were to keep introducing new backend functionalities without harming frontend altought it becomes outdated. And taking into account the situation, I would keep the flyway changes as they're right now (or if you prefer we can remove it from application-template.yml), creating an issue in Taiga/Github to avoid forgeting to revert the change in the production environment.
> 
> Sorry for the extension of the comment. I'm available to hold a remote meeting at any time.

@mariogarcia If you don't have nothing more to add I'll proceed to merge it at the end of the day.
